### PR TITLE
Initial rate limiting implementation

### PR DESCRIPTION
**Pull request by Maxopoly**
This is a follow up to our discussion regarding rate limiting in NameLayer.

It adds easily usable token buckets which can be used for rate limiting or general lag handling (AFKPGC?). It also adds direct integration for commands.

### CODE REVIEW

**Comment by ProgrammerDan** [comment]
would be good to integrate into Tab handling as well if possible, check the namelayer buildout for inspiration.

do you intend this to supersede that implementation, or can they co-exist?

**Comment by Maxopoly** [comment]
They can coexist fine, duplicating useful stuff into CivModCore has long tradition after all. 
NameLayer even still uses it own CommandHandler implementation which was later copied into CivModCore.

It's one of those things that can maybe be cleaned up eventually and doesn't hurt anyone.

**Comment by Maxopoly** [comment]
The command handling rework didn't really make sense in its own PR as long as this isn't merged, so I added it here. This implements what we discussed in Discord, configuration of additional command parameters via plugin.yml and automated binding of implementation through the use of an annotation.

I tested it with a dummy plugin and all of it seemed to work fine

Example plugin.yml with all options:

```
commands:
  yeptest:
    description: Does something
    usage: yeptest a a
    rate-limiter:
      capacity: 5
      amount: 2
      interval: 10s
    tab-rate-limiter:
      capacity: 5
      amount: 2
      interval: 10s
    player-only: false
    console-only: true
    min-args: 1
    max-args: 2
```

and all you need for the command is 

```
@CivCommand(id = "yeptest")
public class TestCommand extends StandaloneCommand {

	@Override
	public boolean execute(CommandSender sender, String[] args) {
		return true;
	}

	@Override
	public List<String> tabComplete(CommandSender sender, String[] args) {
		return null;
	}
}
```

Legacy behavior is priorized to avoid breaking anything existing, but was deprecated. Plugins using this API must extend ACivMod and call super.onEnable or manually instanciate a StandaloneCommandHandler and call it in `onCommand` and `onTabComplete`.

File: src/main/java/vg/civcraft/mc/civmodcore/ratelimiting/TokenBucket.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	 * @param initialCapacity Initial amount of tokens
+	 */
+	TokenBucket(int initialCapacity) {
+		this.tokens = initialCapacity;
+		this.lastRefill = System.currentTimeMillis();
+	}
+
+	public synchronized void refill(int maxTokens, int toAdd, long minDelay) {
+		int refillMultiplier = (int) ((System.currentTimeMillis() - lastRefill) / minDelay);
+		tokens = Math.max(maxTokens, tokens + toAdd * refillMultiplier);
```
Comment (ProgrammerDan) [inline src/main/java/vg/civcraft/mc/civmodcore/ratelimiting/TokenBucket.java]: Do we reset lastRefill ever?

File: src/main/resources/META-INF/services/javax.annotation.processing.Processor [hunk 2]
```diff
@@ -0,0 +1,2 @@
+vg.civcraft.mc.civmodcore.command.CivConfigAnnotationProcessor
```
Comment (ProgrammerDan) [inline src/main/resources/META-INF/services/javax.annotation.processing.Processor]: Neat stuff.

**Comment by Maxopoly** [comment]
Both fixed

**Comment by Maxopoly** [comment]
Should be good to merge

**Pull request merged**